salt/minion.py

```diff
@@ -1424,12 +1424,16 @@ def _thread_return(cls, minion_instance, opts, data):
         function_name = data['fun']
         if function_name in minion_instance.functions:
             try:
-                if minion_instance.connected and minion_instance.opts['pillar'].get('minion_blackout', False):
+                if ((minion_instance.connected and
+                        minion_instance.opts['pillar'].get('minion_blackout', False)) or
+                        minion_instance.opts['grains'].get('minion_blackout', False)):
                     # this minion is blacked out. Only allow saltutil.refresh_pillar
+                    minion_blackout_whitelist = minion_instance.opts['pillar'].get('minion_blackout_whitelist',
+                            minion_instance.opts['grains'].get('minion_blackout_whitelist', []))
```

Comment:
It looks like there are cases when this code will be executed, yet `pillar` is not in `minion_instance.opts`. This case would cause an exception. One such case is when the salt-minion is run and has a configuration such as:

```
master_type: disable
startup_states: top
file_client: local
```

So I would suggest a solution such as:

```
minion_blackout_whitelist = minion_instance.opts.get('pillar', {}).get('minion_blackout_whitelist',
        minion_instance.opts['grains'].get('minion_blackout_whitelist', []))
```